File: src/handlers/formats/exif.js
import { compositeGeoTags, parseCompositeGeo } from "../composite.js";
import { normalizeDiskDate } from "../../core/datetime.js";

export const exifHandler = {
  readTags: [
    ...compositeGeoTags,
    "Composite:GPSDateTime",
    "EXIF:DateTimeOriginal",
    "EXIF:CreateDate",
    "EXIF:ModifyDate",
    "File:FileCreateDate",
    "File:FileModifyDate",
  ],

  parse(raw) {
    const dates = [
      raw["EXIF:DateTimeOriginal"],
      raw["EXIF:CreateDate"],
      raw["Composite:GPSDateTime"],
      raw["EXIF:ModifyDate"],
      raw["File:FileCreateDate"],
      raw["File:FileModifyDate"],
    ];
    return {
      timestamp: normalizeDiskDate(raw["EXIF:DateTimeOriginal"]),
      dates: dates.map(normalizeDiskDate).filter(Boolean),
      geo: parseCompositeGeo(raw),
    };
  },

  timestamp(data) {
    // We use SubSecDateTimeOriginal to update both DateTimeOriginal and OffsetTimeOriginal in one shot.
    return [`SubSecDateTimeOriginal=${data}`];
  },

  geo(data) {
    // For some reason, exiftool does not support writing into Composite:GPSAltitude.
    // Hence, we need to write into GPSAltitude and GPSAltitudeRef separately. GPSAltitudeRef automatically
    // assigns correct value based on the passed value of altitude. For example: if > 0, it will
    // write 0 (above sea level), if < 0, it will write 1 (below sea level).
    return [
      `Composite:GPSLatitude=${data.latitude}`,
      `Composite:GPSLongitude=${data.longitude}`,
      `GPSAltitude=${data.altitude}`,
      `GPSAltitudeRef=${data.altitude}`,
    ];
  },
};

File: packages/core/src/pixelkasten/handlers/exif.py
"""Metadata handler for EXIF-based formats (JPEG, HEIC, PNG)."""

from pixelkasten.core.datetime import normalize_disk_date
from pixelkasten.handlers.shared import (
    COMPOSITE_GEO_TAGS,
    parse_composite_geo,
)


class ExifHandler:
    """Metadata handler for EXIF-based formats (JPEG, HEIC, PNG)."""

    read_tags: list[str] = [
        *COMPOSITE_GEO_TAGS,
        "Composite:GPSDateTime",
        "EXIF:DateTimeOriginal",
        "EXIF:CreateDate",
        "EXIF:ModifyDate",
        "File:FileCreateDate",
        "File:FileModifyDate",
    ]

    _DATE_KEYS = [
        "EXIF:DateTimeOriginal",
        "EXIF:CreateDate",
        "Composite:GPSDateTime",
        "EXIF:ModifyDate",
        "File:FileCreateDate",
        "File:FileModifyDate",
    ]

    def parse(self, raw: dict) -> dict:
        dates = [
            d for d in (normalize_disk_date(raw.get(k)) for k in self._DATE_KEYS) if d is not None
        ]

        return {
            "timestamp": normalize_disk_date(raw.get("EXIF:DateTimeOriginal")),
            "dates": dates,
            "geo": parse_composite_geo(raw),
        }

    def timestamp(self, data: str) -> list[str]:
        return [f"SubSecDateTimeOriginal={data}"]

    def geo(self, data: dict) -> list[str]:
        return [
            f"Composite:GPSLatitude={data['latitude']}",
            f"Composite:GPSLongitude={data['longitude']}",
            f"GPSAltitude={data['altitude']}",
            f"GPSAltitudeRef={data['altitude']}",
        ]
